### PR TITLE
CONSOLE-3164: Move main nav components

### PR DESCRIPTION
**Pull request by raspbeep**
https://issues.redhat.com/browse/CONSOLE-3164

Signed-off-by: Pavel Kratochvil <pakratoc@redhat.com>

### CODE REVIEW

File: frontend/public/locales/en/public.json [hunk 1]
```diff
@@ -1200,16 +1200,6 @@
...
-  "Remove from navigation?": "Remove from navigation?",
-  "Are you sure you want to remove <1>{{label}}</1> from navigation?": "Are you sure you want to remove <1>{{label}}</1> from navigation?",
-  "Nav": "Nav",
-  "All Clusters": "All Clusters",
-  "Advanced Cluster Management": "Advanced Cluster Management",
-  "Navigation": "Navigation",
-  "Pinned resources": "Pinned resources",
-  "Main navigation": "Main navigation",
-  "Drag to reorder": "Drag to reorder",
-  "Unpin": "Unpin",
```
[the remark this answers is not in the document]
Comment (raspbeep) [reply frontend/public/locales/en/public.json:1212]: Fixed, thanks.

File: frontend/packages/console-app/src/components/nav/utils.ts [hunk 2]
```diff
@@ -25,12 +25,14 @@ const findIndexForItem = (item: NavExtension, currentItems: NavExtension[]): num
   const after = toArray(insertAfter);
   let count = 0;
   while (count < before.length && index < 0) {
-    index = currentItems.findIndex((i) => i.properties.id === before[count]);
+    const loopCount = count;
```
Comment (TheRealJon) [inline frontend/packages/console-app/src/components/nav/utils.ts:28]: IMO, we should strongly consider refactoring this function. Adding this const puts a bandaid on the lint issue, but the larger issue is that this function is not adhering to functional programming principles. I have some ideas on how we can improve this if you want to talk more about it.

File: frontend/packages/console-app/src/components/nav/NavHeader.tsx [hunk 3]
```diff
@@ -1,22 +1,22 @@
...
+import { detectFeatures, clearSSARFlags } from '@console/internal/actions/features';
+import { formatNamespaceRoute } from '@console/internal/actions/ui';
+import { history } from '@console/internal/components/utils';
 import { ConsoleLinkModel } from '@console/internal/models';
-import * as acmIcon from '../../imgs/ACM-icon.svg';
+import { K8sResourceKind, referenceForModel } from '@console/internal/module/k8s';
+import { useExtensions } from '@console/plugin-sdk';
+import { useActiveCluster, useActiveNamespace, ACM_LINK_ID } from '@console/shared';
+import { useTelemetry } from '@console/shared/src/hooks/useTelemetry';
+import * as acmIcon from '../../../../../public/imgs/ACM-icon.svg';
```
Comment (TheRealJon) [inline frontend/packages/console-app/src/components/nav/NavHeader.tsx:19]:
```suggestion
import * as acmIcon from '@console/internal/imgs/ACM-icon.svg';
```

File: frontend/packages/console-app/src/components/nav/NavHeader.scss [hunk 4]
```diff
@@ -1,5 +1,5 @@
 @import '~@patternfly/patternfly/sass-utilities/all';
-@import '../../style/vars';
+@import '../../../../../public/style/vars';
```
Comment (TheRealJon) [inline frontend/packages/console-app/src/components/nav/NavHeader.scss:2]: I think we can import style sheets similarly to the way we import js packages:
```suggestion
@import '~@console/internal/style/vars';
```

**Review by TheRealJon — COMMENTED** [review]
Just a few small comments

File: frontend/packages/console-app/src/components/nav/NavHeader.scss (hunk 4 above)
Comment (raspbeep) [reply frontend/packages/console-app/src/components/nav/NavHeader.scss]: Thanks, fixed.

File: frontend/packages/console-app/src/components/nav/NavHeader.tsx (hunk 3 above)
Comment (raspbeep) [reply frontend/packages/console-app/src/components/nav/NavHeader.tsx]: Thanks, fixed.

File: frontend/packages/console-app/src/components/nav/utils.ts [hunk 5]
```diff
@@ -18,25 +18,37 @@ const itemDependsOnItem = (s1: NavExtension, s2: NavExtension): boolean => {
...
+        return newIndex;
+      }
+    }
+    return index;
+  }, -1);
+};
+
+const findInsertAfterIndex = (item: NavExtension, allItems: NavExtension[]): number => {
+  return toArray(item.properties.insertAfter).reduce((index, currentItem) => {
+    if (allItems.findIndex((i) => i.properties.id === currentItem) >= 0) {
```
Comment (TheRealJon) [inline frontend/packages/console-app/src/components/nav/utils.ts:36]:
```suggestion
    if (index < 0) {
```

File: frontend/packages/console-app/src/components/nav/utils.ts [hunk 6]
```diff
@@ -18,25 +18,37 @@ const itemDependsOnItem = (s1: NavExtension, s2: NavExtension): boolean => {
...
-    if (index >= 0) {
-      index += 1;
+const findInsertBeforeIndex = (item: NavExtension, allItems: NavExtension[]): number => {
+  return toArray(item.properties.insertBefore).reduce((index, currentItem) => {
+    // take only the first index found
+    if (index < 0) {
+      const newIndex = allItems.findIndex((i) => i.properties.id === currentItem);
+      if (newIndex !== -1) {
+        return newIndex;
+      }
```
Comment (TheRealJon) [inline frontend/packages/console-app/src/components/nav/utils.ts:28]: Technically we don't need the extra `if` here. We can just return the result of findIndex.
```suggestion
      return allItems.findIndex((i) => i.properties.id === currentItem);
```

**Review by TheRealJon — COMMENTED** [review]
A couple of minor adjustments

File: frontend/packages/console-app/src/components/nav/utils.ts (hunk 5 above)
Comment (raspbeep) [reply frontend/packages/console-app/src/components/nav/utils.ts]: Fixed.

**Comment by jhadvig** [comment]
@raspbeep change looks good 👍 

/lgtm
/approve

QE Approver:
 /assign [@yapei](https://github.com/yapei)
Docs Approver:
 /assign [@opayne1](https://github.com/opayne1)
PX Approver:
 /assign [@RickJWagner](https://github.com/RickJWagner)

**Review by jhadvig — APPROVED** [review]

**Comment by TheRealJon** [comment]
QE Approver
/assign @yapei

Docs Approver:
/assign @opayne1

PX Approver:
/assign @RickJWagner

**Comment by TheRealJon** [comment]
@jhadvig For some reason the `/assign` commands didn't work in your comment, so I repeated them.

**Comment by yapei** [comment]
QE run a regression testing of nav menus, no issues found
- adding/removal of nav menus
- pin/unpin of nav menus
- nav menu highlights etc
/label qe-approved

**Pull request merged**